src/docs/wisp-options.md
Common Wisp user options
---------------------------

This document outlines the common configuration options supported by
  the Wisp core.

Overview
========

The users configurations are generally stored in a "Choices" file and
  are loaded at browser startup. Most user interfaces provide a way
  to configure these parameters in a manner consistant with the
  toolkit in use.

The user choices are stored as a simple key:value list.

Each entry has a type, one of: boolean, integer, unsigned integer,
  hexadecimal colour value or string.

General Options
===============

 Option Key           | Type   | Default   | Description                      
 -------------------- | ------ | --------- | -------------------------------- 
 http_proxy           | bool   | false     | An HTTP proxy should be used.    
 http_proxy           | bool   | false     | An HTTP proxy should be used.    
 http_proxy_host      | string | NULL      | Hostname of proxy.               
 http_proxy_port      | int    | 8080      | Proxy port.                      
 http_proxy_auth      | int    | 0         | Proxy authentication method.     
 http_proxy_auth_user | string | NULL      | Proxy authentication user name   
 http_proxy_auth_pass | string | NULL      | Proxy authentication password    
 http_proxy_noproxy   | string | localhost | Proxy omission list              
 font_size            | int    | 128       | Default font size / 0.1pt.       
 font_min_size        | int    | 85        | Minimum font size.               
 font_sans            | string | NULL      | Default sans serif font          
 font_serif           | string | NULL      | Default serif font               
 font_mono            | string |  NULL     | Default monospace font           
 font_cursive         | string |  NULL     | Default cursive font             
 font_fantasy         | string |  NULL     | Default fantasy font             
 accept_language      | string |  NULL     | Accept-Language header.          
 accept_charset       | string |  NULL     | Accept-Charset header.           
 memory_cache_size    | int    | 12MiB     | Preferred maximum size of memory cache in bytes. 
 disc_cache_size      | uint   | 1GiB      | Preferred expiry size of disc cache in bytes. 
 disc_cache_age       | int    | 28        | Preferred expiry age of disc cache in days. 
 disc_cache_path      | string |  NULL     | Path to disc cache, NULL means to use system path |
 block_advertisements | bool   | false     | Whether to block advertisements  
 do_not_track         | bool   | false     | Disable website tracking [1]     
 send_referer         | bool   | true      | Whether to send the referer HTTP header.
 foreground_images    | bool   | true      | Whether to fetch foreground images 
 background_images    | bool   | true      | Whether to fetch background images 
 animate_images       | bool   | true      | Whether to animate images        
 enable_javascript    | bool   | false     | Whether to execute javascript    
 script_timeout       | int    | 10        | Maximum time to wait for a script to run in seconds 
 expire_url           | int    | 28        | How many days to retain URL data for. 
 font_default         | int    | 0         | Default font family              
 ca_bundle            | string | NULL      | ca-bundle location               
 ca_path              | string | NULL      | ca-path location                 
 cookie_file          | string | NULL      | Cookie file location             
 cookie_jar           | string | NULL      | Cookie jar location              
 homepage_url         | string | NULL      | Home page location               
 search_url_bar       | bool   | false     | search web from url bar          
 search_provider      | int    | 0         | default web search provider      
 url_suggestion       | bool   | true      | URL completion in url bar        
 window_x             | int    | 0         | default x position of new windows 
 window_y             | int    | 0         | default y position of new windows 
 window_width         | int    | 0         | default width of new windows     
 window_height        | int    | 0         | default height of new windows    
 window_screen_width  | int    | 0         | width of screen when above options were saved 
 window_screen_height | int    | 0         | height of screen when above options were saved 
 toolbar_status_size  | int    | 6667      | default size of status bar vs. h scroll bar 
 scale                | int    | 100       | default window scale             
 incremental_reflow   | bool   | true      | Whether to reflow web pages while objects are fetching 
 min_reflow_period    | uint   | 25        | Minimum time (in cs) between HTML reflows while objects are fetching 
 core_select_menu     | bool   | false     | Use core selection menu          

[1] http://www.w3.org/Submission/2011/SUBM-web-tracking-protection-20110224/#dnt-uas

Fetcher options
===============

 Option Key               | Type | Default | Description                         
 ------------------------ | -----| ------- | ----------------------------------- 
 max_fetchers             | int  | 24      | Maximum simultaneous active fetchers 
 max_fetchers_per_host    | int  | 5       | Maximum simultaneous active fetchers per host. (<=option_max_fetchers else it makes no sense) [2]       
 max_cached_fetch_handles | int  |  6      | Maximum number of inactive fetchers cached. The total number of handles wisp will therefore have open is this plus option_max_fetchers. 
 suppress_curl_debug      | bool | true    | Suppress debug output from cURL.    
 target_blank             | bool | true    | Whether to allow target="_blank"    
 button_2_tab             | bool | true    | Whether second mouse button opens in new tab. 

[2] Note that rfc2616 section 8.1.4 says that there should be no more
     than two keepalive connections per host. None of the main browsers
     follow this as it slows page fetches down considerably.
     See https://bugzilla.mozilla.org/show_bug.cgi?id=423377#c4


PDF / Print options
===================

 Option Key             | Type | Deflt | Description                         
 ---------------------- | ---- | ----- | ------------------------------------
 margin_top             | int  | 10    | top margin of exported page          
 margin_bottom          | int  | 10    | bottom margin of exported page       
 margin_left            | int  | 10    | left margin of exported page         
 margin_right           | int  | 10    | right margin of exported page        
 export_scale           | int  | 70    | scale of exported content            
 suppress_images        | bool | false | suppressing images in printed content
 remove_backgrounds     | bool | false | turning off all backgrounds for printed content 
 enable_loosening       | bool | true  | turning on content loosening for printed content 
 enable_PDF_compression | bool | true  | compression of PDF documents         
 enable_PDF_password    | bool | false | setting a password and encoding PDF documents 

System colours
==============

These are the css system colours which the browser also uses to style
generated output.

 Option Key                     | Type   | Default   
 ------------------------------ | ------ | ----------
 sys_colour_ActiveBorder        | colour | 0x00d3d3d3 
 sys_colour_ActiveCaption       | colour | 0x00f1f1f1 
 sys_colour_AppWorkspace        | colour | 0x00f1f1f1 
 sys_colour_Background          | colour | 0x006e6e6e 
 sys_colour_ButtonFace          | colour | 0x00f9f9f9 
 sys_colour_ButtonHighlight     | colour | 0x00ffffff 
 sys_colour_ButtonShadow        | colour | 0x00aeaeae 
 sys_colour_ButtonText          | colour | 0x004c4c4c 
 sys_colour_CaptionText         | colour | 0x004c4c4c 
 sys_colour_GrayText            | colour | 0x00505050 
 sys_colour_Highlight           | colour | 0x00c00800 
 sys_colour_HighlightText       | colour | 0x00ffffff 
 sys_colour_InactiveBorder      | colour | 0x00f1f1f1 
 sys_colour_InactiveCaption     | colour | 0x00e6e6e6 
 sys_colour_InactiveCaptionText | colour | 0x00a6a6a6 
 sys_colour_InfoBackground      | colour | 0x008fdfef 
 sys_colour_InfoText            | colour | 0x00000000 
 sys_colour_Menu                | colour | 0x00f1f1f1 
 sys_colour_MenuText            | colour | 0x004e4e4e 
 sys_colour_Scrollbar           | colour | 0x00cccccc 
 sys_colour_ThreeDDarkShadow    | colour | 0x00aeaeae 
 sys_colour_ThreeDFace          | colour | 0x00f9f9f9 
 sys_colour_ThreeDHighlight     | colour | 0x00ffffff 
 sys_colour_ThreeDLightShadow   | colour | 0x00ffffff 
 sys_colour_ThreeDShadow        | colour | 0x00d5d5d5 
 sys_colour_Window              | colour | 0x00f1f1f1 
 sys_colour_WindowFrame         | colour | 0x004e4e4e 
 sys_colour_WindowText          | colour | 0x00000000 


ChangeLog
16 - Sep 14 2023
- Build fixes oriented at *BSD
- Fixed compilation from a different directory

15 - Aug 21 2023
- Build and runtime fixes for byacc and libressl
- Other build system improvements
- Superior CMake error handling
- Support for older versions of LibreSSL
- Improved usage of libpsl

14 - Jun 11 2023
- Added Wiby search engine
- Fixed AppImages not loading certain libraries
- libpng, libjpeg-turbo and libwebp are now optional
- Critical build system improvements and fixes
- Versioning scheme change
- Added versions to SONAME

0.13 - Jun 10 2023
- CMake prevents building on OpenSSL 1.x.x
- Separated frontends from libwisp source tree
- Renamed "wisp-gtk3" to "wisp-gtk"
- Headers are now installed for libcss, libdom, etc
- Entering an address into the bar uses the current tab instead of opening a
new one
- Merged libdom-hubbub-bindings with libdom
- Various small code improvements
- Introduced basic AppImage support

0.12 - Jun 5 2023
- Minor compatibility improvements
- Added proper check for pkg-config or pkgconf
- Consolidated the search and URL bars on GTK frontend
- Temporarily removed Gtk3 preferences menu; it was broken in upstream
Wisp and should be rewritten

0.11 - May 26 2023
- Added byacc support
- Removed some previous bison artifacts
- Fixed building with clang/lld
- Removed excess warnings
- Added pkg_config checks for m4, zlib and libssl
- Various build system improvements
- Fixed hard-coded pathnames in libwisp, wisp-gtk3 and wisp-vi

0.10 - May 14 2023
- Disabled urlfilter in Visurf
- Visurf now better handles local file paths
- Fixed parser_wrapper.sh failing on some systems
- Added all commands and bindings to README.md
- Added flex and bison as requirements to README.md

0.9 - May 11 2023
- Reworked README.md
- Adding some styling guidelines to CONTRIBUTING.md
- Added new pictures of Wisp
- Improved how cmake handles CFLAGS
- Fixed dependency searching
- Visurf now defaults to about:welcome
- Various minor changes
- Removed about:bindings and about:Choices (in favor of about:config)

0.8 - May 10 2023
- Improved clarity of build instructions in README
- Visurf and Gtk3 frontends now optional
- Resources required for the frontends are now installed properly in
${CMAKE_INSTALL_PREFIX}/share
- libwisp headers are now installed
- CMAKE_INSTALL_LIBDIR variable is now supported

0.7 - May 9 2023
- Added screen of Wisp to README.md
- Added LICENSE file
- Prevent building if source or build dirs have whitespace
- Added concise error messages when build-essential utilities are missing
- Added stub CONTRIBUTING.md file

0.6 - May 8 2023
- Incorporated upstream commits to Wisp and its subprojects
- Improved LibreSSL support
- Fixed frequent or constant crashing on certain websites
- Send DNT (Do Not Track) request by default
- Updated Visurf for compatibility with latest Wisp commits
- Further reductions of unused or obsolete code
- Fixed search engine being set to Dictionary.com instead of DuckDuckGo
- Removed redundant resources
- Changes 'q' binding for closing Visurf to ctrl+q
- Large number of other fixes and improvements

0.5 - May 7 2023
- Began bulk rebranding to Wisp
- Removed support for languages besides for English (these may be
re-introduced later)
- Reworked welcome.html, license.html, and credits.html
- Removed some obsolete documentation
- Removed libnslog
- Merged libwapcaplet into libnsutils
- Enabled JavaScript by default
- Further miscellaneous code reductions
- Replaced libnspsl in favor of a system-installed libpsl

0.4 - May 3 2023
- Switched default browser to DuckDuckGo
- duktape bindings are now generated automatically
- Added a proper install rule
- Added CMake option for installing nsgenbind
- Added CMake option for building and installing an xxd implementation
- Removed librsvg support in favor of libsvgtiny
- Removed support for RiscOS sprites (image/x-riscos-sprite)
- Further reductions to unnecessary source code
- libsvgtiny colors are now generated automatically with gperf
- libcss element properties and selectors are now generated automatically at configure time
- Fixed flex and bison support for nsgenbind
- Added CMake option for installing genparser from libcss

0.3 - Apr 19 2023
- Removed unused libutf8proc support
- Fixed LibreSSL compatibility
- Removes all frontends except for gtk3 and visurf
- Converted build system to CMake
- Removed the about:testament page
- Separated common code into libwisp

0.2 - Apr 13 2023
- Added byacc compatibility to the build system
- Removed the broken and largely useless framebuffer driver
- Minor build system improvements and fixes
- Removed outdated RiscOS target and its associated libraries

0.1 - Mar 12, 2023
- Added Visurf support
- Updated libcss to master-02232023
- Removed Perl from wisp/tools/Makefile
- Disabled some sections of code which break LibreSSL
- Added pre-built text assets to the source tree to bypass Perl dependency
- Updated all library makefiles to build shared instead of static libraries


include/wisp/desktop/options.h
/*
 * Copyright 2012 Vincent Sanders <vince@netsurf-browser.org>
 *
 * This file is part of NetSurf, http://www.netsurf-browser.org/
 *
 * NetSurf is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * NetSurf is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * \file
 * Option available on all platforms
 *
 * Non-platform specific options can be added by editing this file
 *
 * Platform specific options should be added in the platform options.h.
 *
 * This header is specificaly intented to be included multiple times
 *   with different macro definitions so there is no guard
 */

#ifndef WISP_DESKTOP_OPTIONS_H_
#define WISP_DESKTOP_OPTIONS_H_

#include "wisp/types.h"

/* defines for system colour table */
#define NSOPTION_SYS_COLOUR_START NSOPTION_sys_colour_AccentColor
#define NSOPTION_SYS_COLOUR_END NSOPTION_sys_colour_VisitedText

#endif

/** An HTTP proxy should be used. */
NSOPTION_BOOL(http_proxy, false)

/** Hostname of proxy. */
NSOPTION_STRING(http_proxy_host, NULL)

/** Proxy port. */
NSOPTION_INTEGER(http_proxy_port, 8080)

/** Proxy authentication method. */
NSOPTION_INTEGER(http_proxy_auth, OPTION_HTTP_PROXY_AUTH_NONE)

/** Proxy authentication user name */
NSOPTION_STRING(http_proxy_auth_user, NULL)

/** Proxy authentication password */
NSOPTION_STRING(http_proxy_auth_pass, NULL)

/** Proxy omission list */
NSOPTION_STRING(http_proxy_noproxy, "localhost")

/** Default treeview font size (unit: 0.1pt) */
NSOPTION_INTEGER(treeview_font_size, 110)

/** Default font size (unit: 0.1pt) */
NSOPTION_INTEGER(font_size, 128)

/** Minimum font size. */
NSOPTION_INTEGER(font_min_size, 85)

/** Default sans serif font */
NSOPTION_STRING(font_sans, NULL)
/** Default serif font */
NSOPTION_STRING(font_serif, NULL)

/** Default monospace font */
NSOPTION_STRING(font_mono, NULL)

/** Default cursive font */
NSOPTION_STRING(font_cursive, NULL)

/** Default fantasy font */
NSOPTION_STRING(font_fantasy, NULL)

/** Accept-Language header. */
NSOPTION_STRING(accept_language, NULL)

/** Accept-Charset header. */
NSOPTION_STRING(accept_charset, NULL)

/** Preferred maximum size of memory cache / bytes. */
NSOPTION_INTEGER(memory_cache_size, 12 * 1024 * 1024)

/** Preferred location of disc cache, or NULL for system provided location */
NSOPTION_STRING(disc_cache_path, NULL)

/** Preferred expiry size of disc cache / bytes. */
NSOPTION_UINT(disc_cache_size, 1024 * 1024 * 1024)

/** Preferred expiry age of disc cache / days. */
NSOPTION_INTEGER(disc_cache_age, 28)

/** Whether to block advertisements */
NSOPTION_BOOL(block_advertisements, false)

/** Control popups e.g. noniteractive opening windows with window.open */
NSOPTION_BOOL(disable_popups, false)

/** Disable website tracking, see
 * http://www.w3.org/Submission/2011/SUBM-web-tracking-protection-20110224/#dnt-uas
 */
NSOPTION_BOOL(do_not_track, false)

/** Whether to send the referer HTTP header */
NSOPTION_BOOL(send_referer, true)

/** Whether to fetch foreground images */
NSOPTION_BOOL(foreground_images, true)

/** Whether to fetch background images */
NSOPTION_BOOL(background_images, true)

/** Whether to animate images */
NSOPTION_BOOL(animate_images, true)

/** Whether to execute javascript */
NSOPTION_BOOL(enable_javascript, true)

/** Whether to allow Author level CSS. */
NSOPTION_BOOL(author_level_css, true)

/** Maximum time (in seconds) to wait for a script to run */
NSOPTION_INTEGER(script_timeout, 10)

/** How many days to retain URL data for */
NSOPTION_INTEGER(expire_url, 28)

/** Default font family */
NSOPTION_INTEGER(font_default, PLOT_FONT_FAMILY_SANS_SERIF)

/** ca-bundle location */
NSOPTION_STRING(ca_bundle, NULL)

/** ca-path location */
NSOPTION_STRING(ca_path, NULL)

/** Cookie file location */
NSOPTION_STRING(cookie_file, NULL)

/** Cookie jar location */
NSOPTION_STRING(cookie_jar, NULL)

/** Home page location */
NSOPTION_STRING(homepage_url, NULL)

/** search web from url bar */
NSOPTION_BOOL(search_url_bar, true)

/** default web search provider */
NSOPTION_STRING(search_web_provider, "DuckDuckGo")

/** URL completion in url bar */
NSOPTION_BOOL(url_suggestion, true)

/** default x position of new windows */
NSOPTION_INTEGER(window_x, 0)

/** default y position of new windows */
NSOPTION_INTEGER(window_y, 0)

/** default width of new windows */
NSOPTION_INTEGER(window_width, 0)

/** default height of new windows */
NSOPTION_INTEGER(window_height, 0)

/** default size of status bar vs. h scroll bar */
NSOPTION_INTEGER(toolbar_status_size, 6667)

/** default window scale */
NSOPTION_INTEGER(scale, 100)

/* Whether to reflow web pages while objects are fetching */
/* incremental_reflow option removed - use CMake -DNEOSURF_ENABLE_INCREMENTAL_REFLOW=ON instead */

/* Minimum time (in cs) between HTML reflows while objects are fetching */
NSOPTION_UINT(min_reflow_period, DEFAULT_REFLOW_PERIOD)

/* use core selection menu */
NSOPTION_BOOL(core_select_menu, false)

/* display decoded international domain names */
NSOPTION_BOOL(display_decoded_idn, false)

/******** Fetcher options ********/

/** Maximum simultaneous active fetchers */
NSOPTION_INTEGER(max_fetchers, 256)

/** Maximum simultaneous active fetchers per host.
 *
 * Note that rfc2616 section 8.1.4 says that there should be no more
 * than two keepalive connections per host. None of the main browsers
 * follow this as it slows page fetches down considerably.  See
 * https://bugzilla.mozilla.org/show_bug.cgi?id=423377#c4
 */
NSOPTION_INTEGER(max_fetchers_per_host, 64)

/** Maximum number of inactive fetchers cached.  The total number of
 * handles netsurf will therefore have open is this plus
 * option_max_fetchers.
 */
NSOPTION_INTEGER(max_cached_fetch_handles, 6)

/** Number of times to retry timed-out fetches before giving up. */
NSOPTION_UINT(max_retried_fetches, 1)

/** Number of seconds to allow for a DNS-resolution+connect() before timing out
 * the cURL socket.
 */
NSOPTION_UINT(curl_fetch_timeout, 30)

/** Suppress debug output from cURL. */
NSOPTION_BOOL(suppress_curl_debug, true)

/******** appearnce of new browser views ********/

/** Whether to allow target="_blank" */
NSOPTION_BOOL(target_blank, true)

/** auxillary (second) mouse button opens in new tab */
NSOPTION_BOOL(button_2_tab, true)

/** control if new view (window or tab) opens in foreground */
NSOPTION_BOOL(foreground_new, false)

/******** PDF / Print options ********/

/** top margin of exported page */
NSOPTION_INTEGER(margin_top, DEFAULT_MARGIN_TOP_MM)

/** bottom margin of exported page */
NSOPTION_INTEGER(margin_bottom, DEFAULT_MARGIN_BOTTOM_MM)

/** left margin of exported page */
NSOPTION_INTEGER(margin_left, DEFAULT_MARGIN_LEFT_MM)

/** right margin of exported page */
NSOPTION_INTEGER(margin_right, DEFAULT_MARGIN_RIGHT_MM)

/** scale of exported content */
NSOPTION_INTEGER(export_scale, DEFAULT_EXPORT_SCALE * 100)

/** suppressing images in printed content */
NSOPTION_BOOL(suppress_images, false)

/** turning off all backgrounds for printed content */
NSOPTION_BOOL(remove_backgrounds, false)

/** turning on content loosening for printed content */
NSOPTION_BOOL(enable_loosening, true)

/** compression of PDF documents */
NSOPTION_BOOL(enable_PDF_compression, true)

/** setting a password and encoding PDF documents */
NSOPTION_BOOL(enable_PDF_password, false)

/** whether to prefer dark mode (light on dark) */
NSOPTION_BOOL(prefer_dark_mode, false)

/******** System colours ********/
NSOPTION_COLOUR(sys_colour_AccentColor, 0x00666666)
NSOPTION_COLOUR(sys_colour_AccentColorText, 0x00ffffff)
NSOPTION_COLOUR(sys_colour_ActiveText, 0x000000ee)
NSOPTION_COLOUR(sys_colour_ButtonBorder, 0x004e4e4e)
NSOPTION_COLOUR(sys_colour_ButtonFace, 0x00f9f9f9)
NSOPTION_COLOUR(sys_colour_ButtonText, 0x004c4c4c)
NSOPTION_COLOUR(sys_colour_Canvas, 0x00f1f1f1)
NSOPTION_COLOUR(sys_colour_CanvasText, 0x00000000)
NSOPTION_COLOUR(sys_colour_Field, 0x00f1f1f1)
NSOPTION_COLOUR(sys_colour_FieldText, 0x00000000)
NSOPTION_COLOUR(sys_colour_GrayText, 0x00a6a6a6)
NSOPTION_COLOUR(sys_colour_Highlight, 0x00c00800)
NSOPTION_COLOUR(sys_colour_HighlightText, 0x00ffffff)
NSOPTION_COLOUR(sys_colour_LinkText, 0x00ee0000)
NSOPTION_COLOUR(sys_colour_Mark, 0x0000ffff)
NSOPTION_COLOUR(sys_colour_MarkText, 0x00000000)
NSOPTION_COLOUR(sys_colour_SelectedItem, 0x00e48435)
NSOPTION_COLOUR(sys_colour_SelectedItemText, 0x00ffffff)
NSOPTION_COLOUR(sys_colour_VisitedText, 0x008b1a55)

/** Filter for non-verbose logging */
NSOPTION_STRING(log_filter, WISP_BUILTIN_LOG_FILTER)
/** Filter for verbose logging */
NSOPTION_STRING(verbose_filter, WISP_BUILTIN_VERBOSE_FILTER)

/* page colour selection */
NSOPTION_UINT(colour_selection, 0)
